virttest/qemu_devices/qcontainer.py

```diff
@@ -1720,6 +1720,150 @@ def images_define_by_params(self, name, image_params, media=None,
+        devices[-1].format_params(devices[-1].params)
+
+        # For virtio serial devices, generate controller and create the port device
+        if serial_type.startswith('virtio'):
+            if not bus:
+                if serial_type == 'virtio-serial-device':
+                    pci_bus = {'type': 'virtio-bus'}
+                else:
+                    pci_bus = {'aobject': 'pci.0'}
+                bus = self.get_first_free_bus(
+                    {'type': 'SERIAL', 'atype': serial_type}, [None, nr])
+                if bus is None:
+                    _hba = serial_type.replace('-', '_') + '%s'
+                    bus = self.idx_of_next_named_bus(_hba)
+                    bus = self.list_missing_named_buses(_hba, 'SERIAL', bus + 1)[-1]
+                    logging.debug("list missing named bus: %s", bus)
+                    devices.append(
+                        qdevices.QDevice(serial_type,
+                                         {"id": bus},
+                                         bus,
+                                         pci_bus,
+                                         qdevices.QSerialBus(bus, serial_type, bus)))
+                else:
+                    bus = bus.busid
+            devices.append(
+                qdevices.QDevice(port_type,
+                                 {"id": serial_id},
+                                 parent_bus={'busid': bus}))
+            devices[-1].set_param('name', serial_name)
+        else:  # none virtio type, generate serial device directly
+            devices.append(qdevices.QDevice(serial_type, {"id": serial_id}))
+            devices[-1].set_param("reg", reg)
+        devices[-1].set_param('chardev', chardev_id)
+
+        return devices
+
+    def serials_define_by_params(self, serial_id, serial_params,
+                                 file_name=None):
+        """
+        Wrapper for creating serial device from serial params.
+
+        :param serial_id: id of serial object
+        :param serial_params: serial params
+        :param file_name: the file path of the serial device (optional)
+        :return: the device list that construct the serial device
+        """
+        serial_type = serial_params['serial_type']
+        machine = serial_params.get('machine_type')
+
+        # Arm lists "isa-serial" as supported but can't use it,
+        # fallback to "-serial"
+        legacy_cmd = " -serial unix:'%s',server,nowait" % file_name
+        legacy_dev = qdevices.QStringDevice('SER-%s' % serial_id,
+                                            cmdline=legacy_cmd)
+        arm_serial = (serial_type == 'isa-serial'
+                      and 'arm' in serial_params.get("machine_type", ""))
+        if (arm_serial or not self.has_option("chardev")
+            or not self.has_device(serial_type)):
+            return legacy_dev
+
+        virtio_port_type = serial_params.get('virtio_port_type')
+        if virtio_port_type:
```

Comment:
In qemu-vm, we already use:
```
            virtio_port_type = serial_params.get('virtio_port_type')
            if virtio_port_type:
                serial_params['serial_type'] = 'virtio-serial'
```
and have got `serial_type` from above, so how about  `if serial_type.startswith('virtio')`  or `if serial_type == 'virtio-serial'` ?   and i prefer the second one.

Reply:
@yanan-fu And for here, qemu-vm is not the only user for the API, we must make sure `virtio_port_type` is defined for `virtio-serial` type anytime when user call it, e.g. define a new device for hotplug.

Reply:
@vivianQizhu  I have different option,  the `serial_type` is depends on `virtio_port_type` all the time with the current implement, in this API,  it already use `L1798    serial_type = serial_params['serial_type']`.

I am not very clear about other usage now, but if it is a problem,  i think we need to update `serial_type` first.